exchanges/exchange.go

```diff
@@ -1136,3 +1113,112 @@ func (e *Base) SetSaveTradeDataStatus(enabled bool) {
 		log.Debugf(log.Trade, "Set %v 'SaveTradeData' to %v", e.Name, enabled)
 	}
 }
+
+// NewEndpoints declares default and running URLs maps
+func (e *Base) NewEndpoints() *Endpoints {
+	return &Endpoints{
+		defaults: make(map[string]string),
+	}
+}
+
+// SetDefaultEndpoints declares and sets the default URLs map
+func (e *Endpoints) SetDefaultEndpoints(m map[URL]string) {
+	for k, v := range m {
+		e.defaults[k.String()] = v
+	}
+}
+
+// SetRunning populates running URLs map
+func (e *Endpoints) SetRunning(key, val string, overwrite bool) error {
+	e.Lock()
+	defer e.Unlock()
+	_, err := url.ParseRequestURI(val)
+	if err != nil {
+		return fmt.Errorf("invalid url: %v", err)
+	}
+	if !overwrite {
+		oldVal, ok := e.defaults[key]
+		if ok && oldVal == val {
+			return fmt.Errorf("given key and val are already set")
+		}
+	}
+	err = validateKey(key)
+	if err != nil {
+		return err
+	}
```

Comment:
This should go right after locking+defer unlock. Better to fail early that someone is trying to set an invalid key, rather than after everything has been calculated and setup

Reply:
This one still isn't done. I mean like this:
```
e.Lock()
	defer e.Unlock()
err = validateKey(key)
	if err != nil {
		return err
	}
	_, err := url.ParseRequestURI(val)
...
```
So we check that the key exists before even trying to set a URL